# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/IonBuilder.cpp
# Commit: 95a4904b7465
# Full Hash: 95a4904b746585a09c96f16820b559bc8c17c996
# Author: Jan de Mooij <jdemooij@mozilla.com>
# Date: 2019-11-26 16:16:58
# Regressor Bug: 1599117
# File Overlap Count: 1
# Description:
#   Bug 1599117 part 1 - Remove createFallthroughJoinBlock, create the join block upfront if needed. r=tcampbell
#   
#   createFallthroughJoinBlock was necessary at some point, but it no longer is
#   because we know before the loop whether we have incoming edges.
#   
# ==============================================================================

diff -r c7a6f863a0cc -r 95a4904b7465 js/src/jit/IonBuilder.cpp
--- a/js/src/jit/IonBuilder.cpp	Sat Nov 23 23:30:36 2019 +0000
+++ b/js/src/jit/IonBuilder.cpp	Tue Nov 26 11:27:18 2019 +0000
@@ -3485,20 +3485,23 @@
   PendingEdges edges(std::move(p->value()));
   pendingEdges_->remove(p);
 
+  // Loop-restarts may clear the list rather than remove the map entry entirely.
+  // This is to reduce allocator churn since it is likely the list will be
+  // filled in again in the general case.
+  if (edges.empty()) {
+    return Ok();
+  }
+
   MBasicBlock* joinBlock = nullptr;
 
-  auto createFallthroughJoinBlock = [this, &joinBlock]() -> AbortReasonOr<Ok> {
-    MOZ_ASSERT(!joinBlock);
+  // Create join block if there's fall-through from the previous bytecode op.
+  if (!hasTerminatedBlock()) {
     MOZ_TRY_VAR(joinBlock, newBlock(current, pc));
     current->end(MGoto::New(alloc(), joinBlock));
     setTerminatedBlock();
-    return Ok();
-  };
+  }
 
   auto addEdge = [&](MBasicBlock* pred, size_t popped) -> AbortReasonOr<Ok> {
-    if (!joinBlock && !hasTerminatedBlock()) {
-      MOZ_TRY(createFallthroughJoinBlock());
-    }
     if (joinBlock) {
       if (!joinBlock->addPredecessorPopN(alloc(), pred, popped)) {
         return abort(AbortReason::Alloc);
@@ -3515,10 +3518,7 @@
   // pass can optimize.
   auto createBlockForShortCircuit =
       [&](MBasicBlock* pred) -> AbortReasonOr<MBasicBlock*> {
-    if (!joinBlock) {
-      MOZ_ASSERT(!hasTerminatedBlock());
-      MOZ_TRY(createFallthroughJoinBlock());
-    }
+    MOZ_ASSERT(joinBlock);
 
     MBasicBlock* trueBlock;
     MOZ_TRY_VAR(trueBlock, newBlock(pred, pc));
@@ -3578,10 +3578,7 @@
     MOZ_CRASH("Invalid kind");
   }
 
-  if (!joinBlock) {
-    return Ok();
-  }
-
+  MOZ_ASSERT(joinBlock);
   MOZ_TRY(startTraversingBlock(joinBlock));
 
   // If the join block has just one predecessor with an MTest, try to improve
